web/pdf_print_service.js

```diff
@@ -0,0 +1,289 @@
+      }
+    },
+
+    destroy: function () {
+      this.printContainer.textContent = '';
+      if (this.pageStyleSheet && this.pageStyleSheet.parentNode) {
+        this.pageStyleSheet.parentNode.removeChild(this.pageStyleSheet);
+        this.pageStyleSheet = null;
+      }
+      if (activeService === this) {
+        activeService = null;
+      }
+      if (scratchCanvas) {
+        scratchCanvas.width = scratchCanvas.height = 0;
+        scratchCanvas = null;
+      }
+      ensureOverlay().then(function () {
+        OverlayManager.close('printServiceOverlay');
+      });
+    },
+
+    renderPages: function () {
+      var pageCount = this.pagesOverview.length;
+      var renderNextPage = function (resolve, reject) {
+        if (++this.currentPage >= pageCount) {
+          renderProgress(pageCount, pageCount);
+          resolve();
+          return;
+        }
+        var index = this.currentPage;
+        renderProgress(index, pageCount);
+        renderPage(this.pdfDocument, index + 1,
+                   this.pagesOverview[index], this.wrappers[index]).then(
+          function () { renderNextPage(resolve, reject); }, reject);
+      }.bind(this);
+      return new Promise(renderNextPage);
+    },
+  };
+
+
+  var print = window.print;
+  window.print = function print() {
+    if (activeService) {
+      console.warn('Ignored window.print() because of a pending print job.');
+      return;
+    }
+    ensureOverlay().then(function () {
+      OverlayManager.open('printServiceOverlay');
+    });
+
+    try {
+      dispatchEvent('beforeprint');
+    } finally {
+      if (!activeService) {
+        console.error('Expected print service to the initialized.');
+      }
+      activeService.renderPages().then(startPrint, abort);
+    }
+  };
+
+  function dispatchEvent(eventType) {
+    var event = document.createEvent('CustomEvent');
+    event.initCustomEvent(eventType, false, false, 'custom');
+    window.dispatchEvent(event);
+  }
+
+  function startPrint() {
+    // Push window.print in the macrotask queue to avoid being affected by
+    // the deprecation of running print() code in a microtask, see
+    // https://github.com/mozilla/pdf.js/issues/7547.
+    setTimeout(function() {
+      if (!activeService) {
+        return; // Print task cancelled by user.
+      }
+      print.call(window);
+      setTimeout(abort, 20); // Tidy-up
+    }, 0);
+  }
+
+  function abort() {
+    if (activeService) {
+      activeService.destroy();
+      dispatchEvent('afterprint');
+    }
+  }
+
+  function renderProgress(index, total) {
+    var progressContainer = document.getElementById('printServiceOverlay');
+    var progress = Math.round(100 * index / total);
+    var progressBar = progressContainer.querySelector('progress');
+    var progressPerc = progressContainer.querySelector('.relative-progress');
+    progressBar.value = progress;
+    progressPerc.textContent = progress + '%';
+    progressContainer.onclick = abort;
```

Comment:
As mentioned on IRC, I think that the `onclick` handler should be attached to the `printCancel` button rather than the *entire* overlay container.